File: languages/fortran/tests/position_tape_tests.f90
program position_tape_tests
  use position_tape
  implicit none

  type(validation_result) :: result
  type(mismatch) :: found
  type(hash_window_index) :: index_result
  integer, allocatable :: positions(:)
  character(len=:), allocatable :: tape, fragment
  character(len=64) :: hash

  call check(generate(0) == "", "zero length")
  call check(generate(11) == "12345678911", "basic generation")
  call check(len(generate(100)) == 100, "exact length")
  call check(len(generate_marker_complete(100)) == 101, "marker complete 100")
  call check(len(generate_marker_complete(10000)) == 10003, "marker complete 10000")

  result = validate(generate(250), 250)
  call check(result%is_valid, "valid tape")

  result = validate(generate(40), 50)
  call check(.not. result%is_valid, "truncated invalid")
  call check(result%truncation_point == 41, "truncation point")

  found = find_first_mismatch(generate(20), generate(19) // "X")
  call check(found%has_value .and. found%position == 20, "mismatch")

  tape = generate(80)
  fragment = tape(30:41)
  call check(locate(fragment) == 30, "locate")
  hash = hash_fragment(fragment)
  index_result = build_window_index(len(fragment))
  call check(any(index_result%hashes == hash .and. index_result%positions == 30), "hash index")
  positions = locate_by_hash(hash, len(fragment))
  call check(any(positions == 30), "locate by hash")

  print *, "OK fortran"

contains
  subroutine check(condition, message)
    logical, intent(in) :: condition
    character(len=*), intent(in) :: message
    if (.not. condition) error stop message
  end subroutine check
end program position_tape_tests

File: languages/ocaml/tests/position_tape_tests.ml
#directory "languages/ocaml/src";;
#use "position_tape.ml";;

let assert_true condition message =
  if not condition then (prerr_endline ("FAIL: " ^ message); exit 1)
;;

assert_true (generate 0 = "") "zero length";;
assert_true (generate 11 = "12345678911") "basic generation";;
assert_true (String.length (generate 100) = 100) "exact boundary";;
assert_true (String.length (generate_marker_complete 100) = 101) "marker complete 100";;
assert_true (String.length (generate_marker_complete 10000) = 10003) "marker complete 10000";;

let valid = validate (generate 250) 250;;
assert_true valid.is_valid "valid tape";;

let truncated = validate (generate 40) 50;;
assert_true (not truncated.is_valid) "truncated invalid";;
assert_true (truncated.truncation_point = Some 41) "truncation point";;

let mutated = Bytes.of_string (generate 60);;
Bytes.set mutated 19 'X';;
let mismatch = find_first_mismatch (generate 60) (Bytes.to_string mutated);;
assert_true ((Option.get mismatch).position = 20) "first mismatch";;

assert_true (find_truncation_point (generate 75) = 76) "find truncation";;
let fragment = String.sub (generate 80) 29 12;;
assert_true (locate fragment = 30) "locate fragment";;
assert_true
  (hash_fragment "" = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855")
  "sha256 empty";;
assert_true
  (hash_fragment "abc" = "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad")
  "sha256 abc";;
let hash = hash_fragment fragment;;
let index = build_window_index (String.length fragment);;
assert_true (List.mem 30 (Hashtbl.find index hash)) "hash index";;
assert_true (List.mem 30 (locate_by_hash (String.uppercase_ascii hash) (String.length fragment))) "locate by hash";;

print_endline "OK ocaml";;
